### PR TITLE
Fix bug that did not remove an unfunded offer (RIPD-1298)

### PR DESCRIPTION
**Pull request by seelabs**
@scottschurr @nbougalis

### CODE REVIEW

File: src/ripple/app/paths/impl/BookStep.cpp [hunk 1]
```diff
@@ -412,7 +412,8 @@ BookStep<TIn, TOut>::revImp (
             result.in = sum(savedIns);
             result.out = out;
             this->consumeOffer (sb, offer, ofrAdjAmt, stpAdjAmt, ownerGivesAdj);
-            return false;
+            // Offers with tiny amounts may be fully consumed even remainingOut < stpAmt.out
```
Comment (vinniefalco) [inline src/ripple/app/paths/impl/BookStep.cpp]: Grammar?

**Comment by vinniefalco** [comment]
We need a much better explanation of this bug fix in the commit message.

**Comment by scottschurr** [comment]
Three thoughts:

- Do we need a comparable fix in `BookStep::fwdImp()`?  The  specific case I ran into happened in `revImp()`, so I'm just guessing that it could happen in  `fwdImp()`.  I'm thinking of something around line 565 that would look like...
```
return processMore || offer.fully_consumed();
```
- This change will affect transaction outcomes, so it will cause de-syncing among servers running with vs without this fix.  Should we save the fix and bundle it with another that takes an amendment?  Or should we put this on a time switch?  
- It would be great to see a unit test for this.  Perhaps one test for the `revImp()` case and another test for the `fwdImp()` case?

**Comment by JoelKatz** [comment]
This will need to be hooked to an amendment or something since it changes transaction processing behavior.

**Comment by vinniefalco** [comment]
Message should read:
```
Fix unfunded offer not removed:

(Description)
```

**Comment by scottschurr** [comment]
Looks magnificent.  :+1:

**Comment by seelabs** [comment]
@vinniefalco Will fix the title when I squash it down.

**Comment by seelabs** [comment]
@vinniefalco (Description will be what's in the second commit now)

File: src/ripple/app/paths/impl/BookStep.cpp [hunk 2]
```diff
@@ -412,7 +412,15 @@ BookStep<TIn, TOut>::revImp (
             result.in = sum(savedIns);
             result.out = out;
             this->consumeOffer (sb, offer, ofrAdjAmt, stpAdjAmt, ownerGivesAdj);
-            return false;
+
+            // When the mantissas of two iou amounts differ by less than ten, then
+            // subtracting them leaves a result of zero. This can cause the check for
+            // (stpAmt.out > remainingOut) to incorrectly think an offer will be funded
+            // after subtracting remainingIn.
```
Comment (vinniefalco) [inline src/ripple/app/paths/impl/BookStep.cpp]: Excellent comment, this is what I need

File: src/test/app/Flow_test.cpp [hunk 3]
```diff
@@ -1305,6 +1305,80 @@ struct Flow_test : public beast::unit_test::suite
...
+
+            auto const alice = Account("alice");
+            auto const bob = Account("bob");
+            auto const gw = Account("gw");
+            auto const USD = gw["USD"];
+
+            env.fund(XRP(100000), alice, bob, gw);
+            env(trust(alice, USD(20)));
+
+            STAmount tinyAmt1(USD.issue(), 9000000000000000ll, -17, false,
```
Comment (vinniefalco) [inline src/test/app/Flow_test.cpp]: nit: braced-init

Comment (seelabs) [inline src/test/app/Flow_test.cpp]: Fixed

**Comment by seelabs** [comment]
Squashed and rebased onto 0.40.0-b3

File: src/ripple/app/paths/impl/BookStep.cpp [hunk 4]
```diff
@@ -561,6 +569,14 @@ BookStep<TIn, TOut>::fwdImp (
 
         remainingIn = in - result.in;
         this->consumeOffer (sb, offer, ofrAdjAmt, stpAdjAmt, ownerGivesAdj);
+
+        // When the mantissas of two iou amounts differ by less than ten, then
+        // subtracting them leaves a result of zero. This can cause the check for
+        // (stpAmt.in > remainingIn) to incorrectly think an offer will be funded
+        // after subtracting remainingIn.
+        if (amendmentRIPD1298(sb.parentCloseTime()))
+            processMore = processMore || offer.fully_consumed();
```
Comment (seelabs) [inline src/ripple/app/paths/impl/BookStep.cpp]: `|=` is not short-circuited when `processMore` is already true. We'll save a function call.

File: src/ripple/ledger/View.h [hunk 5]
```diff
@@ -347,6 +347,12 @@ bool amendmentRIPD1274 (NetClock::time_point const closeTime);
 NetClock::time_point const& amendmentRIPD1274SoTime ();
 bool amendmentRIPD1274 (NetClock::time_point const closeTime);
 
+NetClock::time_point const& amendmentRIPD1298SoTime ();
+bool amendmentRIPD1298 (NetClock::time_point const closeTime);
+
+NetClock::time_point const& amendmentRIPD1298SoTime ();
+bool amendmentRIPD1298 (NetClock::time_point const closeTime);
```
Comment (seelabs) [inline src/ripple/ledger/View.h]: I just noticed I've been doing that for every single switch-over. Fixed.

File: src/ripple/app/paths/impl/BookStep.cpp (hunk 4 above)
Comment (nbougalis) [inline src/ripple/app/paths/impl/BookStep.cpp]: Fair enough.

**Comment by scottschurr** [comment]
Just to be clear, still :+1:.

**Comment by seelabs** [comment]
Squashed

**Comment by nbougalis** [comment]
Merged as bb0b97f46bcc783955e0fcd1511fa2086c48cb1a

**Pull request closed without merging**